File: Eratosthenes/Eratosthenes.py
def eratosthenes(n):
# This function returns a list of prime numbers up to n using the Sieve of Eratosthenes algorithm.
# The algorithm works by marking the multiples of each prime number starting from 2 as non-prime.
# The remaining numbers are prime.
    primeNumber = 2
    primeNumbers = [True] * (n+1)
    while primeNumber * primeNumber < n:
        if primeNumbers[primeNumber]:
            for i in range(primeNumber*primeNumber, n+1, primeNumber):
                primeNumbers[i] = False
        primeNumber += 1
# the reason using primeNumber * primeNumber is when we reach a prime number p, all multiples of p less than p^2 have already been marked as non-prime
    return [i for i in range(2, n+1) if primeNumbers[i]]
        
# Example usage:
n = int(input("Enter the number to find prime numbers up to it: "))
print(f"Prime numbers up to {n}: {eratosthenes(n)}")

File: Eratosthenes/Eratosthenes.cpp
#include<iostream>
#include<vector>
using namespace std;

vector<int> eratosthenes(int n);
vector<int> eratosthenesRecursive(int n);

int main()
{
    int n;
    cout<<"Enter the number upto which you want to find the prime numbers: ";
    cin>>n;
    vector<int> primes = eratosthenes(n);
    cout<<"The prime numbers upto "<<n<<" are: ";
    for(int prime : primes) {
        cout << prime << " ";
    }
    cout << endl;
    return 0;
}

// Function to find all prime numbers upto n using Eratosthenes algorithm
vector<int> eratosthenes(int n){
    vector<bool> primeNumbers(n+1, true);
    vector<int> primeNumbersList;
    int currentPrimeNumber = 2;
    while (currentPrimeNumber*currentPrimeNumber <= n){
        if (primeNumbers[currentPrimeNumber]){
            for(int i = currentPrimeNumber*currentPrimeNumber; i <= n; i+= currentPrimeNumber)
                primeNumbers[i] = false;
        }
        currentPrimeNumber++;
    }
    for(int i = 2; i <= n; i++){
        if (primeNumbers[i])
            primeNumbersList.push_back(i);
    }
    return primeNumbersList;
}

// Function to find all prime numbers upto n using Eratosthenes algorithm recursively
vector<int> eratosthenesRecursion(int n, int currentPrimeNumber, vector<bool>& primeNumbers) {
    if (currentPrimeNumber * currentPrimeNumber > n) {
        vector<int> primeNumbersList;
        for (int i = 2; i <= n; i++) {
            if (primeNumbers[i]) {
                primeNumbersList.push_back(i);
            }
        }
        return primeNumbersList;
    }

    if (primeNumbers[currentPrimeNumber]) {
        for (int i = currentPrimeNumber * currentPrimeNumber; i <= n; i += currentPrimeNumber) {
            primeNumbers[i] = false;
        }
    }

    return eratosthenesRecursion(n, currentPrimeNumber + 1, primeNumbers);
}

vector<int> eratosthenesRecursive(int n) {
    vector<bool> primeNumbers(n + 1, true);
    return eratosthenesRecursion(n, 2, primeNumbers);
}